### ISSUE TITLE
Convenience Function to get Distkey & Sortkey

### ISSUE THREAD

**Issue by elyse-weiss** [move-coop/parsons#364]
Given the new warnings in the Redshift class, I would find it super useful to have a convenience function that gets you distkey and sortkey.

**Comment by shaunagm** [comment]
Is this something we still want? If so, can you provide more information, @elyse-weiss?

**Comment by elyse-weiss** [comment]
Yes! Distkey and Sortkeys are very important and Redshift, and I do think this would be a used convenience function - especially if you need to drop and re-create a table.

**Comment by shaunagm** [comment]
Okay, bumping up the priority. Does TMC have a redshift sandbox that a contributor could access while working on this?

**Comment by elyse-weiss** [comment]
Upon request, I could definitely create a Redshift user and give them access to our `communal_data_assets` schema, which just includes publicly available data, so they can play around there!

**Comment by HandcartCactus** [comment]
This seems to be stalled because it requires Redshift (a paid service). I've got an AWS account. I could adopt this ticket, if it's worthwhile.

**Comment by shaunagm** [comment]
@HandcartCactus that would be amazing!

**Comment by HandcartCactus** [comment]
Sounds good. This is my first issue so it might take a while to get spun up.

This issue doesn't state what the actual warning is or when it comes up in the Redshift class, so I'll try to see if I can get it to appear first.

From my research online, a table's distkey and sortkey can be found in svv_table_info and/or pg_table_def if the table is in the user's search_path.

**Comment by HandcartCactus** [comment]
Huh. I'm trying to use the Redshift class for the first time and it appears to be "invisible" to my language server/mypy. I've tried this both installed directly from github and the distribution from pypi. I wish the DX was better. 

<img width="366" height="78" alt="Image" src="https://github.com/user-attachments/assets/03963383-ec13-4b70-aea6-61aaadd59e7f" />

This appears to be caused by imports coming from the `__init__.py` `_CONNECTORS` list as strings, instead of as regular import statements. Here's the imports:
<img width="665" height="700" alt="Image" src="https://github.com/user-attachments/assets/ab6b3d21-4f5e-441c-ab9c-6d66a33c34fc" />
Here's how they get imported, which is quite interesting.
<img width="727" height="485" alt="Image" src="https://github.com/user-attachments/assets/7261ac9e-abe1-4871-bc4b-080a1b27377b" />

I see that you're doing this to get around the issue of various modules having "optional" dependencies (which I personally also find to be a headache). Optional dependencies are kind of unavoidable in an all-encompassing package like this; but it shouldn't have to completely break developer support tools. For example, I think `pandas` has an optional dependency issue without breaking DX. It requires openpyxl or xlrd if you're doing excel writing, but it only ever comes up if you try using them, and tools like mypy work just fine. 

<img width="1458" height="657" alt="Image" src="https://github.com/user-attachments/assets/9a2176fb-02b5-4f3c-bb82-93f60e5c838e" />

<img width="702" height="375" alt="Image" src="https://github.com/user-attachments/assets/1677644c-d328-42cf-b46e-427f8c39b699" />

Pandas uses a combination of tricks to accomplish this (conditional imports, imports within method calls, factory-like functions, `import_optional_dependency`). 

Once I get through this, I would love to explore what could be done in Parsons to improve the DX.

**Comment by HandcartCactus** [comment]
Looks like someone attempted to improve this issue in #1810 but it's not working for me.

**Comment by HandcartCactus** [comment]
<img width="672" height="165" alt="Image" src="https://github.com/user-attachments/assets/b3716c0d-f973-4e29-8a73-ecd22e231bd3" />

Using full direct path instead of import from top level package (as shown in tutorial) works. Lesson learned :)

**Comment by shaunagm** [comment]
_Once I get through this, I would love to explore what could be done in Parsons to improve the DX._

Whenever you're ready, please reach out and we can do a 1:1. And/or if you're free Thurs the 2nd at 2pm ET, we'll have our next contributor meeting then.

Are you in the Slack, by the way?

**Comment by HandcartCactus** [comment]
> Are you in the Slack, by the way?

I am! Just sent you a message.

**Comment by HandcartCactus** [comment]
@shaunagm 

1. Can I add typehints to method signatures? I've done it, but it seems to be rare. Is that ok?
2. All methods return a Table (or None?). Should they be something else? Table might be overkill, but it's also a common return type.
3. Should a missing table return None, raise an exception, or should that be configurable (e.g. `errors='raise'`)? It currently returns None.
4. Is it required for my docstrings to be RST? I see a mix of styles and markup languages in this class.
5. Do the tests happen in a real redshift instance?
You can see my current progress here:
https://github.com/move-coop/parsons/compare/main...HandcartCactus:parsons:feature/364-redshift-distkey-sortkey

**Comment by shaunagm** [comment]
1. Yes please do add type hints! We like them, we just haven't prioritized adding them.
2. Most Parsons methods return tables but they don't have to if it simply doesn't make sense. Maybe we can ask what people prefer in the Parsons Slack? Otherwise I think either is fine.
3. I like the idea of configuring whether an error is raised or not. Maybe by default return None but allow people to raise an exception instead?
4. Ideally the docstring is rst, we should probably do an audit of the docstrings that are not, for consistency's sake.
5. For testing, we typically do a combination of mocks, which are run on CI/CD, and live tests on real third-party instances, which are typically only run when a person is actively editing the connector. If there's a free or very cheap way to run the live tests as part of CI that'd be great, but so far none of our connectors do that - instead you use the live test decorator and toggle it on when running live test for the connector you're specifically working on.

**Comment by HandcartCactus** [comment]
Almost done! I've learned a lot about redshift distkeys and sortkeys.

**Comment by shaunagm** [comment]
Amazing! I'm a bit OOO because I'm at a conference this week but feel free to ping me for review whenever you're ready, and we can also chat about it next week at the contributor meeting (thurs 2pm ET) if you're available then

**Comment by HandcartCactus** [comment]
Thanks @shaunagm! I think it's ready for review. Is that something we can do asynchronously, or do you need me virtually present? I'm usually offline 9am-7pm Mon-Fri but I do have some flexibility there.

**Comment by shaunagm** [comment]
We should be able to do it async! I'll take a look at this ithin the next day or two.

**Comment by shaunagm** [comment]
Closing as completed, thank you @HandcartCactus!
